numpy/_core/src/common/npy_argparse.h

```diff
@@ -4,6 +4,13 @@
 #include <Python.h>
 #include "numpy/ndarraytypes.h"
 
+#ifdef __STDC_NO_ATOMICS__
+#define atomic_int volatile int
```

Comment:
Caveat: this is "good enough" for x86 MSVC and also Window's "ARM64EC" platform target, but not other ARM64 targets on Windows.

If we need to support other Windows ARM64 targets, we'll need a bit more work. Hopefully, MSVC will support `<stdatomic.h>` soon, but I wouldn't hold my breath for it.

https://learn.microsoft.com/en-us/cpp/build/reference/volatile-volatile-keyword-interpretation

Reply:
Thanks for clarifying! We don't test or build wheels for windows ARM64 right now, but I think we I think we might need to support `_M_ARM64`. According to the MSVC docs, `ARM64EC` sets `_M_64` I see `pyatomic_msc.h` has:

```
static inline uint64_t
_Py_atomic_load_uint64(const uint64_t *obj)
{
#if defined(_M_X64) || defined(_M_IX86)
    return *(volatile uint64_t *)obj;
#elif defined(_M_ARM64)
    return (uint64_t)__ldar64((unsigned __int64 volatile *)obj);
#else
#  error "no implementation of _Py_atomic_load_uint64"
#endif
}
```

And storing is implemented using e.g. `_InterlockedCompareExchange64` from `intrin.h`.

That all seems reasonable to include in NumPy to support this.

Even with the experimental atomics support in MSVC 2022, they still set `_ STDC_NO_ATOMICS _`, so we'd need to go out of our way to detect and use it.